### DISCUSSION TITLE
Optimisation of point update in loop

### DISCUSSION THREAD

**Discussion by jkissi** [marcomusy/vedo#501] [Q&A] [answered]
Hi! 
I am incorporating vedo as the core visualisation in my research, I think its a great library but I am now coming up against problems of speed. In my case the thing I want to update between viewer refreshes is not the position of the points but _the colour_ of the points.
 
![test_capture_46_edit](https://user-images.githubusercontent.com/8246894/139554243-75b6fe12-919d-4ba5-b573-6f9b9a59279a.png)

My algorithm runs through the point cloud and classifies points into classes using different colours, and I want to see how those classifications evolve by periodically updating the vedo viewer. So, in this situation, the position of the points in constant between refreshes.  I've done a little debugging and found that the line seems to really take a long time during each update cycle is the following:

```
         import vedo as ve
         
         # Initialise the point could 
         pcd = ve.pointcloud.Points(points)
        
         for n in np.arange(0, length_of_loop):
              # algorithmic implementation ....
 
              # this is the line that seems to cause the real slowdown
              pcd = ve.Points(pcd.points(), r=2, c=colour_array)
        
             # more algorithmic implementation ....
             
             # update the loop here 
             vis.show(pcd, lineset_array, text_list, interactive=False, zoom=1.5)

    # show the final result here
    vis.show(pcd, lineset_array, text_list, interactive=True, zoom=1.5)
```

in the above line "colour_array" is a list of RGB triplets, one triplet for each point in the cloud. Is this the best way to update the colors of the cloud? Any suggestions you may have to optimise the speed here would be very gratefully accepted.

Thanks! 
J

**Comment by marcomusy** [comment, marked as answer]
Hi, thanks for your nice words, the problem is that you recreate from scratch the whole `Points` object. I would it this way:
```python
import vedo as ve
import numpy as np

ve.settings.defaultFont = "Theemim"

plt = ve.Plotter(interactive=False)

# Initialise the point could
points = np.random.rand(1000,3)
pcd = ve.pointcloud.Points(points, r=10)
pcd.pointdata["MY_RGB"] = np.zeros([1000,3], dtype=np.uint8)
pcd.pointdata.select("MY_RGB") # can add an alpha channel too

plt += [pcd, ve.Axes(pcd)]
plt.show()

pb = ve.ProgressBar(0, 500)
for n in range(500):
    colour_array = np.random.rand(pcd.NPoints(), 3)
    colour_array /= (np.max(colour_array) / 255) # normalize correctly
    colour_array = colour_array.astype(np.uint8) # this will save memory
    pcd.pointdata["MY_RGB"] = colour_array

    # update the loop here
    plt.render()
    pb.print()

ve.interactive()
```
which runs at about 60Hz on my pc.

**Reply by jkissi** [reply]
Hi Marco 
Thanks for the answer, its very interesting!! But I seem to be having a problem and I have a few questions. First the problem:
When I try your example exactly, it works great, but when attempt to put it into my code I get the following:

![faded_edit](https://user-images.githubusercontent.com/8246894/139751856-13d59308-023b-4539-a2d3-e9154fd565cd.png)

you can see that the lines are still there but the point colours are all messed up. I am very confused about that because the code example you gave seems pretty straight-forward...I checked the colour_array its an array of ints with three colours: red green and white, so I'm at a bit of a loss. Any ideas?

Here are my other questions: 
how does the pcd know that the array relates to the point colour? Is "MY_RGB" a special dictionary for colour, or is that dictionary being **_created_** in the example you gave?
What does the pcd.pointdata.select() function do?

I feel like I must be missing something simple here...

 Cheers,
J

**Reply by marcomusy** [reply]
could it be that you have some `nan` in your list of colors? I see in your screenshot that nan shows up in the text at top-left.

> how does the pcd know that the array relates to the point colour?

an array of uint8 of length 3 or 4 is automatically rendered as rgba color, the name can be whatever you choose. The `select()` selects  the active array to be visualized among the existing arrays.

Go step by step: e.g. give your pointcloud the same random array as in the working example and see if it still works...

**Reply by jkissi** [reply]
Hi Marco!
I took your advice and finally got everything working and sorted out. The update is now much, much faster. I've included a short vid, so you can get an idea of what I've managed to get so far. 

https://user-images.githubusercontent.com/8246894/140099131-2ec3aac5-ac4e-4018-adc4-2ff6f4231c2d.mp4

The gaps between updates are now entirely down to my classifier having to process between update. This is a great position for me to continue from. Can't thank you enough. That select() function is awesome. Is it documented somewhere? I couldn't find it...

**Reply by jkissi** [reply]
Oh the video doesn't display. do you know how to add video on here?

**Reply by marcomusy** [reply]
> That select() function is awesome. Is it documented somewhere? 

Indeed it's a bit hidden:
https://vedo.embl.es/autodocs/content/vedo/base.html?highlight=pointdata#vedo.base.BaseActor.pointdata

> Oh the video doesn't display. do you know how to add video on here?

Yes, I can see it. It looks cool even though I don't know exactly what it is! LOL
I'm very happy that it worked out! Thanks for the feedback, let me know if you need any more help,
Marco
